src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,2089 @@
+                new_circ.add(gates.M(1, basis=gates.Z))
+
+                expectation_val = GST_measure_2qb(new_circ, k, j)
+
+            # print(expectation_val)
+            matrix_jk_2qb[j, k] = expectation_val
+
+    matrix_jk_2qb = np.matrix(matrix_jk_2qb)
+
+    return matrix_jk_2qb
+
+
+def GST_1qb_basis_operations(
+    NshotsGST=1e4, noise_model=None, gjk_1qb=None, save_data=None, backend=None
+):
+    """Runs gate set tomography for the 13 basis operations for a single qubit.
+
+        Args:
+        NshotsGST (int, optional): Number of shots used in Gate Set Tomography.
+        noise_model (:class:`qibo.noise.NoiseModel`, optional): Noise model applied
+            to simulate noisy computation.
+        gjk_1qb (numpy.matrix): Matrix with elements Tr(Q_j rho_k) for one qubit with no gate.
+            If None, call GST_1qb without gate.
+        save_data: Flag to save gate set tomography data. If None, skip.
+        backend (:class:`qibo.backends.abstract.Backend`, optional): Calculation engine.
+
+        The 13 basis operations are based on Physical Review Research 3, 033178 (2021).
+    Returns:
+        numpy.matrix: Matrix with elements jk equivalent to either Tr(Q_j rho_k) or Tr(Q_j O_l rho_k).
+    """
+
+    import numpy as np
+
+    from qibo import gates
+    from qibo.backends import GlobalBackend
+    from qibo.models import Circuit
+
+    if backend is None:  # pragma: no cover
+        backend = GlobalBackend()
+
+    if gjk_1qb is None:
+        gjk_1qb = GST_1qb(NshotsGST, noise_model, save_data=None)
+
+    BasisOps_13 = [
+        np.matrix(gates.I(0).matrix(backend=GlobalBackend())),
+        np.matrix(gates.X(0).matrix(backend=GlobalBackend())),
+        np.matrix(gates.Y(0).matrix(backend=GlobalBackend())),
+        np.matrix(gates.Z(0).matrix(backend=GlobalBackend())),
+        np.matrix(
+            [
+                [1 / np.sqrt(2) * 1, 1 / np.sqrt(2) * 1j],
+                [1 / np.sqrt(2) * 1j, 1 / np.sqrt(2) * 1],
+            ]
+        ),
+        np.matrix(
+            [
+                [1 / np.sqrt(2) * 1, 1 / np.sqrt(2) * 1],
+                [1 / np.sqrt(2) * (-1), 1 / np.sqrt(2) * 1],
+            ]
+        ),
+        np.matrix([[1 / np.sqrt(2) * (1 + 1j), 0], [0, 1 / np.sqrt(2) * (1 - 1j)]]),
+        np.matrix(
+            [
+                [1 / np.sqrt(2) * 1, 1 / np.sqrt(2) * (-1j)],
+                [1 / np.sqrt(2) * 1j, 1 / np.sqrt(2) * (-1)],
+            ]
+        ),
+        np.matrix(
+            [
+                [1 / np.sqrt(2) * 1, 1 / np.sqrt(2) * 1],
+                [1 / np.sqrt(2) * 1, 1 / np.sqrt(2) * (-1)],
+            ]
+        ),
+        np.matrix([[0, 1 / np.sqrt(2) * (1 - 1j)], [1 / np.sqrt(2) * (1 + 1j), 0]]),
+        np.matrix([[1, 0], [0, 0]]),
+        np.matrix([[0.5, 0.5], [0.5, 0.5]]),
+        np.matrix([[0.5, -0.5j], [0.5j, 0.5]]),
+    ]
+
+    def GST_measure_13(circuit, idx_basis_op, k, j):
```

Comment:
All these functions seem to be very similar at a first glance, wouldn't it be possible to generalize them to a unique function that does the things in a slightly different way depending on the inputs?

Reply:
These basis operations are from [https://arxiv.org/pdf/2006.12509.pdf)](https://arxiv.org/pdf/2006.12509.pdf). The first 4 are Pauli, easy to implement using ```np.matrix(gates.I(0).matrix(backend=GlobalBackend()))``` while the rest are combinations of Clifford gates. I've worked them out previously and these are the matrices. In reply to two comments above, yes, the 8th (zero based indexing) is a Hadamard which I will correct to ```np.matrix(gates.H(0).matrix(backend=GlobalBackend()))```.